File: src/protocol/commands/text.ts
/**
 * Text display command builder.
 *
 * Text characters are rendered to 16x32 monochrome bitmaps, packed 8 pixels per byte (LSB first).
 * Characters are separated by 0x05 0xFF 0xFF 0xFF.
 *
 * Packet format:
 *   [16-byte header] + [14-byte metadata] + [bitmap data]
 *
 * Header:
 *   [0-1]:  u16 LE total packet length
 *   [2-4]:  0x03, 0x00, 0x00 (static)
 *   [5-8]:  u32 LE length of metadata + bitmaps
 *   [9-12]: u32 LE CRC32 of metadata + bitmaps
 *   [13-15]: 0x00, 0x00, 0x0C (static)
 *
 * Metadata:
 *   [0-1]:  u16 LE number of characters (separator count)
 *   [2-3]:  0x00, 0x01 (static)
 *   [4]:    text_mode (0-8)
 *   [5]:    speed
 *   [6]:    text_color_mode
 *   [7-9]:  text RGB
 *   [10]:   text_bg_mode
 *   [11-13]: bg RGB
 */

import type { Color, TextBgMode, TextColorMode, TextMode } from "../types.ts";
import { crc32 } from "../../utils/crc32.ts";

const TEXT_SEPARATOR = new Uint8Array([0x05, 0xff, 0xff, 0xff]);
const IMAGE_WIDTH = 16;
const IMAGE_HEIGHT = 32;

export interface TextOptions {
	text: string;
	bitmaps: Uint8Array;
	textMode?: TextMode;
	speed?: number;
	textColorMode?: TextColorMode;
	textColor?: Color;
	textBgMode?: TextBgMode;
	textBgColor?: Color;
}

export function buildTextPacket(opts: TextOptions): Uint8Array {
	const {
		bitmaps,
		textMode = 1,
		speed = 95,
		textColorMode = 1,
		textColor = { r: 255, g: 0, b: 0 },
		textBgMode = 0,
		textBgColor = { r: 0, g: 0, b: 0 },
	} = opts;

	// Count separators in bitmaps to get num_chars
	let numChars = 0;
	for (let i = 0; i <= bitmaps.length - 4; i++) {
		if (
			bitmaps[i] === 0x05 &&
			bitmaps[i + 1] === 0xff &&
			bitmaps[i + 2] === 0xff &&
			bitmaps[i + 3] === 0xff
		) {
			numChars++;
		}
	}

	// Build metadata
	const metadata = new Uint8Array(14);
	const metaView = new DataView(metadata.buffer);
	metaView.setUint16(0, numChars, true);
	metadata[2] = 0x00;
	metadata[3] = 0x01;
	metadata[4] = textMode;
	metadata[5] = speed & 0xff;
	metadata[6] = textColorMode;
	metadata[7] = textColor.r & 0xff;
	metadata[8] = textColor.g & 0xff;
	metadata[9] = textColor.b & 0xff;
	metadata[10] = textBgMode;
	metadata[11] = textBgColor.r & 0xff;
	metadata[12] = textBgColor.g & 0xff;
	metadata[13] = textBgColor.b & 0xff;

	// Combine metadata + bitmaps
	const packet = new Uint8Array(metadata.length + bitmaps.length);
	packet.set(metadata, 0);
	packet.set(bitmaps, metadata.length);

	// Build header
	const header = new Uint8Array(16);
	const headerView = new DataView(header.buffer);
	const totalLen = header.length + packet.length;
	headerView.setUint16(0, totalLen, true);
	header[2] = 0x03;
	header[3] = 0x00;
	header[4] = 0x00;
	headerView.setUint32(5, packet.length, true);
	headerView.setUint32(9, crc32(packet), true);
	header[13] = 0x00;
	header[14] = 0x00;
	header[15] = 0x0c;

	// Combine header + packet
	const result = new Uint8Array(totalLen);
	result.set(header, 0);
	result.set(packet, header.length);
	return result;
}

/**
 * Render a single character to a 16x32 monochrome bitmap.
 * Returns separator + bitmap bytes.
 *
 * Each row: 16 pixels packed into 2 bytes, LSB first within each byte.
 * Total: 32 rows * 2 bytes = 64 bytes per character.
 */
export function renderCharBitmap(pixels: Uint8Array): Uint8Array {
	// pixels should be IMAGE_WIDTH * IMAGE_HEIGHT values of 0 or 1
	const bitmap = new Uint8Array(IMAGE_WIDTH * IMAGE_HEIGHT / 8);
	let byteIndex = 0;
	let currentByte = 0;

	for (let y = 0; y < IMAGE_HEIGHT; y++) {
		for (let x = 0; x < IMAGE_WIDTH; x++) {
			if (x % 8 === 0) {
				currentByte = 0;
			}
			const pixel = pixels[y * IMAGE_WIDTH + x] ?? 0;
			currentByte |= (pixel & 1) << (x % 8);
			if (x % 8 === 7 || x === IMAGE_WIDTH - 1) {
				bitmap[byteIndex++] = currentByte;
			}
		}
	}

	// Prepend separator
	const result = new Uint8Array(TEXT_SEPARATOR.length + bitmap.length);
	result.set(TEXT_SEPARATOR, 0);
	result.set(bitmap, TEXT_SEPARATOR.length);
	return result;
}

File: rust-cli/src/protocol/commands/text.rs
use crate::protocol::types::Color;

const TEXT_SEPARATOR: [u8; 4] = [0x05, 0xFF, 0xFF, 0xFF];
const IMAGE_WIDTH: usize = 16;
const IMAGE_HEIGHT: usize = 32;

pub struct TextOptions {
    pub bitmaps: Vec<u8>,
    pub text_mode: u8,
    pub speed: u8,
    pub text_color_mode: u8,
    pub text_color: Color,
    pub text_bg_mode: u8,
    pub text_bg_color: Color,
}

impl Default for TextOptions {
    fn default() -> Self {
        Self {
            bitmaps: Vec::new(),
            text_mode: 1,
            speed: 95,
            text_color_mode: 1,
            text_color: Color::new(255, 0, 0),
            text_bg_mode: 0,
            text_bg_color: Color::new(0, 0, 0),
        }
    }
}

pub fn build_text_packet(opts: &TextOptions) -> Vec<u8> {
    // Count separators in bitmaps to get num_chars
    let mut num_chars: u16 = 0;
    if opts.bitmaps.len() >= 4 {
        for i in 0..=opts.bitmaps.len() - 4 {
            if opts.bitmaps[i] == 0x05
                && opts.bitmaps[i + 1] == 0xFF
                && opts.bitmaps[i + 2] == 0xFF
                && opts.bitmaps[i + 3] == 0xFF
            {
                num_chars += 1;
            }
        }
    }

    // Build metadata (14 bytes)
    let mut metadata = vec![0u8; 14];
    metadata[0..2].copy_from_slice(&num_chars.to_le_bytes());
    metadata[2] = 0x00;
    metadata[3] = 0x01;
    metadata[4] = opts.text_mode;
    metadata[5] = opts.speed;
    metadata[6] = opts.text_color_mode;
    metadata[7] = opts.text_color.r;
    metadata[8] = opts.text_color.g;
    metadata[9] = opts.text_color.b;
    metadata[10] = opts.text_bg_mode;
    metadata[11] = opts.text_bg_color.r;
    metadata[12] = opts.text_bg_color.g;
    metadata[13] = opts.text_bg_color.b;

    // Combine metadata + bitmaps
    let mut packet = metadata;
    packet.extend_from_slice(&opts.bitmaps);

    // Build header (16 bytes)
    let total_len = 16 + packet.len();
    let packet_crc = crc32fast::hash(&packet);

    let mut header = vec![0u8; 16];
    header[0..2].copy_from_slice(&(total_len as u16).to_le_bytes());
    header[2] = 0x03;
    header[3] = 0x00;
    header[4] = 0x00;
    header[5..9].copy_from_slice(&(packet.len() as u32).to_le_bytes());
    header[9..13].copy_from_slice(&packet_crc.to_le_bytes());
    header[13] = 0x00;
    header[14] = 0x00;
    header[15] = 0x0C;

    // Combine header + packet
    let mut result = header;
    result.extend(packet);
    result
}

/// Render a single character to a 16x32 monochrome bitmap.
/// Returns separator + bitmap bytes.
///
/// Each row: 16 pixels packed into 2 bytes, LSB first within each byte.
/// Total: 32 rows * 2 bytes = 64 bytes per character.
pub fn render_char_bitmap(pixels: &[u8]) -> Vec<u8> {
    let mut bitmap = vec![0u8; IMAGE_WIDTH * IMAGE_HEIGHT / 8];
    let mut byte_index = 0;
    let mut current_byte: u8 = 0;

    for y in 0..IMAGE_HEIGHT {
        for x in 0..IMAGE_WIDTH {
            if x % 8 == 0 {
                current_byte = 0;
            }
            let pixel = pixels.get(y * IMAGE_WIDTH + x).copied().unwrap_or(0);
            current_byte |= (pixel & 1) << (x % 8);
            if x % 8 == 7 || x == IMAGE_WIDTH - 1 {
                bitmap[byte_index] = current_byte;
                byte_index += 1;
            }
        }
    }

    let mut result = Vec::with_capacity(TEXT_SEPARATOR.len() + bitmap.len());
    result.extend_from_slice(&TEXT_SEPARATOR);
    result.extend(bitmap);
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_render_char_bitmap_empty() {
        let pixels = vec![0u8; IMAGE_WIDTH * IMAGE_HEIGHT];
        let result = render_char_bitmap(&pixels);
        assert_eq!(result.len(), 4 + 64); // separator + bitmap
        assert_eq!(&result[0..4], &TEXT_SEPARATOR);
        assert!(result[4..].iter().all(|&b| b == 0));
    }

    #[test]
    fn test_render_char_bitmap_first_pixel() {
        let mut pixels = vec![0u8; IMAGE_WIDTH * IMAGE_HEIGHT];
        pixels[0] = 1; // top-left pixel
        let result = render_char_bitmap(&pixels);
        assert_eq!(result[4], 0x01); // LSB set
    }

    #[test]
    fn test_build_text_packet_header() {
        let bitmaps = render_char_bitmap(&vec![0u8; IMAGE_WIDTH * IMAGE_HEIGHT]);
        let opts = TextOptions {
            bitmaps,
            ..Default::default()
        };
        let result = build_text_packet(&opts);

        // Header checks
        assert_eq!(result[2], 0x03);
        assert_eq!(result[3], 0x00);
        assert_eq!(result[4], 0x00);
        assert_eq!(result[13], 0x00);
        assert_eq!(result[14], 0x00);
        assert_eq!(result[15], 0x0C);

        // Metadata: num_chars = 1
        assert_eq!(result[16], 1);
        assert_eq!(result[17], 0);
    }
}
